src/Http/Routing/src/Constraints/FileNameRouteConstraint.cs

```diff
@@ -0,0 +1,146 @@
+    ///     <item>  
+    ///         <term><c>hello.world.txt</c></term>  
+    ///         <description>Final segment contains a <c>.</c> followed by other characters.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>.gitignore</c></term>  
+    ///         <description>Final segment contains a <c>.</c> followed by other characters.</description>  
+    ///     </item> 
+    /// </list>
+    /// <list type="bullet">  
+    ///     <listheader>  
+    ///         <term>Examples of route values that will be rejected as non-file-names</term>  
+    ///         <description>description</description>  
+    ///     </listheader>  
+    ///     <item>  
+    ///         <term><c>/a/b/c</c></term>  
+    ///         <description>Final segment does not contain a <c>.</c>.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>/a/b.d/c</c></term>  
+    ///         <description>Final segment does not contain a <c>.</c>.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>/a/b.d/c/</c></term>  
+    ///         <description>Final segment is empty.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c></c></term>  
+    ///         <description>Value is empty</description>  
+    ///     </item>
+    /// </list>  
+    /// </para>
+    /// </remarks>
+    public class FileNameRouteConstraint : IRouteConstraint
+    {
+        /// <inheritdoc />
+        public bool Match(
+            HttpContext httpContext,
+            IRouter route,
+            string routeKey,
+            RouteValueDictionary values,
+            RouteDirection routeDirection)
+        {
+            if (routeKey == null)
+            {
+                throw new ArgumentNullException(nameof(routeKey));
+            }
+
+            if (values == null)
+            {
+                throw new ArgumentNullException(nameof(values));
+            }
+
+            if (values.TryGetValue(routeKey, out var obj) && obj != null)
+            {
+                var value = Convert.ToString(obj, CultureInfo.InvariantCulture);
+                return IsFileName(value);
+            }
+
+            // No value or null value.
+            return false;
+        }
+
+        internal static bool IsFileName(ReadOnlySpan<char> value)
+        {
+            if (value.Length == 0)
+            {
+                // Not a file name because empty.
+                return false;
+            }
+
+            var lastSlashIndex = value.LastIndexOf('/');
+            if (lastSlashIndex >= 0)
+            {
+                value = value.Slice(lastSlashIndex + 1);
+            }
+
+            var dotIndex = value.IndexOf('.');
+            if (dotIndex == -1)
+            {
+                // No dot.
+                return false;
+            }
+
+            for (var i = dotIndex + 1; i < value.Length; i++)
+            {
+                if (value[i] != '.')
```

Comment:
I don't super follow this. Wouldn't this prevent `foo.bar.baz` which the documentation says is allowed?

Reply:
It only needs one non-period character after the first period.

Reply:
Hmm, so if I understand correctly, it's trying to disallow files that look like `.....`, while allowing files that look like `..a.b...c.`? This needs some comments.

Reply:
My goal is to allow `foo.bar` and `.foobar` but not nonesense like `..`.